File: basic_dsa_two_hundred_sixteen.cpp
/**
 * Basic DSA - Kth Smallest Element in a Sorted Matrix (LeetCode 378)
 * Find kth smallest element in n x n matrix where each row/col is sorted.
 */
#include <iostream>
#include <vector>
#include <algorithm>

int kthSmallest(std::vector<std::vector<int>>& matrix, int k) {
    int n = matrix.size();
    std::vector<int> flat;
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            flat.push_back(matrix[i][j]);
    std::nth_element(flat.begin(), flat.begin() + k - 1, flat.end());
    return flat[k - 1];
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Kth Smallest Element in a Sorted Matrix\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<std::vector<int>> m1 = {{1, 5, 9}, {10, 11, 13}, {12, 13, 15}};
    std::cout << "[[1,5,9],[10,11,13],[12,13,15]], k=8 -> " << kthSmallest(m1, 8) << "\n";

    std::vector<std::vector<int>> m2 = {{-5}};
    std::cout << "[[-5]], k=1 -> " << kthSmallest(m2, 1) << "\n";
}

File: basic_dsa_two_hundred_sixteen.py
"""
Basic DSA - Find the Duplicate Number
Given an array nums containing n + 1 integers where each integer is in the range
[1, n] inclusive, there is only one repeated number. Return the repeated number.
You must not modify the array and use only constant extra space.
"""


def find_duplicate(nums):
    """
    Floyd's Tortoise and Hare cycle detection.
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    slow = nums[0]
    fast = nums[0]
    while True:
        slow = nums[slow]
        fast = nums[nums[fast]]
        if slow == fast:
            break

    slow2 = nums[0]
    while slow != slow2:
        slow = nums[slow]
        slow2 = nums[slow2]
    return slow


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Find the Duplicate Number")
    print("=" * 60)

    print(find_duplicate([1, 3, 4, 2, 2]))  # 2
    print(find_duplicate([3, 1, 3, 4, 2]))  # 3
